game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/change/RemoveUnprotectedUnitsGeneral.java

```diff
@@ -0,0 +1,22 @@
+package games.strategy.triplea.delegate.battle.steps.change;
+
+import games.strategy.triplea.delegate.battle.BattleActions;
+import games.strategy.triplea.delegate.battle.BattleState;
+import java.util.List;
+
+public class RemoveUnprotectedUnitsGeneral extends RemoveUnprotectedUnits {
+  public RemoveUnprotectedUnitsGeneral(
+      final BattleState battleState, final BattleActions battleActions) {
+    super(battleState, battleActions);
+  }
+
+  @Override
+  public List<String> getNames() {
+    return List.of();
+  }
+
+  @Override
+  public Order getOrder() {
+    return Order.REMOVE_UNPROTECTED_UNITS_GENERAL;
```

Comment:
It does not look like there is much implementation here and this is mostly a configuration class.

What do you think of creating a more generic variant of this class? It looks like all we would need to do it to parameterize the 'getOrder()' value as a constructor arg, and then we could simply return it. Can we potentially reduce the inheritance tree even and have the super class do some of this work, particularly if it had a constructor that accepted the `getNames()` and `getOrder()` values? Since those are constants, it seems like we probably could new-up the parent directly.

Reply:
Looking at this a bit more, I do think we should avoid the inheritance tree and see if we can flatten this. A rule of thumb is to never override methods that have implementations. It generally makes code hard to understand, test, and inheritance is the strongest form of coupling. Instead just make the parent method abstract, convert the parent class into two, one that is abstract and a second that has the original implementation from the parent. Then this class would become a second child.

What that would look like is:
![parent-hierarchy](https://user-images.githubusercontent.com/12397753/93653265-f9ee2580-f9cc-11ea-81b6-fabf4052e02a.png)


But, the differences between the child classes are just constants, configuration, not behavior. In this case I'd flatten it more so you would just new up the BattleStep via a factory type of pattern and inject the right configurations and some behavior for the execution, EG:

```
BattleStep RemoveUnprotectedUnits = BattleStepFactory.builder()
    .names( ... )
    .order( ...)
    .executable( removeUnprotectedUnitsExecutable(bridge))
    .build()
    .toBattleStep();
```

WDYT?

Reply:
The problem is save compatibility.  `RemoveUnprotectedUnits` is part of the saved file and so it has to be able to deserialize correctly.  I can add a parent to it but I can't move the two private variables.  That would mean that I would need to add abstract `getBattleState` and `getBattleActions` methods to the parent class and then have the two child classes overload them to make the variables available to the parent methods.  That just feels weird.

As for the builder idea, that means that now the configuration is part of the saved file and it makes it even harder to change code in the future.  I would much rather keep it in classes with as little class state as possible.

Reply:
I kinda suspect there might be some ways around the serialization issue, though perhaps not.. It is of course frustrating that serialization concerns really impacts our code design (and in this case, forces inheritance. I wonder a bit how many "Java inheritance now considered harmful" articles there are by now).